Allow graph navigation by browser forward/backward

AustinMroz opened the pull request:
- On graph change, set the `graph.id` as location hash
- On hash change, navigate to the target `graph.id` either in the current, or any other loaded workflow.

`canvasStore.currentGraph` does not trigger when `app.loadGraphData` is called. A trigger could be forced here, but I'm concerned about side effects. Instead `updateHash` is manually called.

Code search shows that there are no current custom nodes using `onhashchange`

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-6811-Allow-graph-navigation-by-browser-forward-backward-2b26d73d365081bb8414fdf7c3686124) by [Unito](https://www.unito.io)

AustinMroz committed:
Allow graph navigation by browser forward/backward

christian-byrne commented on src/stores/subgraphNavigationStore.ts:
```diff
@@ -157,6 +156,60 @@ export const useSubgraphNavigationStore = defineStore(
+    //Allow navigation with forward/back buttons
+    //TODO: Extend for dialogues?
+    //TODO: force update widget.promoted
+    function onHashChange() {
```
How about on page reload? I.e., initial load

AustinMroz committed:
Use router for managing location hash

AustinMroz committed:
Fix broken test

christian-byrne commented:
Would consider using https://vueuse.org/router/useRouteHash/

AustinMroz commented:
> Would consider using https://vueuse.org/router/useRouteHash/

Already got pinged by Alex for this and made the change. Will push once I fix your suggestion of making navigation work on initial load.

AustinMroz committed:
Fix breadcrumb bar always navigating to root graph

AustinMroz committed:
Restore subgraph on initial load

DrJKL commented on browser_tests/fixtures/ws.ts:
```diff
@@ -29,7 +29,7 @@ export const webSocketFixture = base.extend<{
               if (!url) {
                 // If no URL specified, use page URL
-                const u = new URL(window.location.toString())
+                const u = new URL(window.location.toString().split('#')[0])
```
Could you use this? https://developer.mozilla.org/en-US/docs/Web/API/Location/hash

DrJKL commented on src/components/breadcrumb/SubgraphBreadcrumbItem.vue:
Should this be a [RouterLink](https://router.vuejs.org/api/variables/RouterLink.html#RouterLink)?

AustinMroz replied:
From quick testing, it wouldn't be a quick substitution. It's causing errors on click about missing state.

Perhaps we move in the opposite direction and replace the `<a>` with a `<div>` since it's not actually performing a navigation?

AustinMroz committed:
Fix incorrect operator precedence

AustinMroz committed:
Fix linting

AustinMroz committed:
Mock out vueuse/router

AustinMroz committed:
Fix first workflow swap skipped

AustinMroz committed:
Do not set location hash until subgraph navigation

AustinMroz committed:
Remove vueuse/router

AustinMroz committed:
Unfix additional test

AustinMroz committed:
Remove app.graph usage

AustinMroz committed:
Navigation cleanup, fix test

Myestery commented on src/stores/subgraphNavigationStore.ts:
```diff
@@ -158,13 +158,87 @@ export const useSubgraphNavigationStore = defineStore(
+    //Allow navigation with forward/back buttons
+    const routeHash = ref(window.location.hash)
+    const originalOnHashChange = window.onhashchange
+    window.onhashchange = (...args) => {
```
we could use `addEventListener` here to avoid overwriting existing handlers

AustinMroz replied:
That's much nicer. :+1:

AustinMroz committed:
use addEventListener

AustinMroz committed:
Fix test

AustinMroz committed:
Fix test

Issues here have been particularly frustrating to work around. I'll need
to find a proper solution at some point

AustinMroz committed:
Reintroduce vue-router

AustinMroz committed:
Small post router tweaks

routeHash updates immediately, so some logic can be simplified and the
newId needs to be pulled prior to initialization

AustinMroz committed:
Fix test

DrJKL reviewed (dismissed):
I'd love an integration test or two, but let's try it 👍🏻

AustinMroz committed:
nit

Co-authored-by: Alexander Brown <drjkl@comfy.org>
